source/mesh_deformation/fastscape.cc

```diff
@@ -0,0 +1,1632 @@
+                  h[i] = h[i] + h_seed;
+                }
+
+              // Here we add the sediment rain (m/yr) as a flat increase in height.
+              // This is done because adding it as an uplift rate would affect the basement.
+              if (sediment_rain > 0 && use_marine)
+                {
+                  // Only apply sediment rain to areas below sea level.
+                  if (h[i] < sl)
+                    {
+                      // If the rain would put us above sea level, set height to sea level.
+                      if (h[i] + sediment_rain*a_dt > sl)
+                        h[i] = sl;
+                      else
+                        h[i] = std::min(sl,h[i] + sediment_rain*a_dt);
+                    }
+                }
+            }
+
+          // The ghost nodes are added as a single layer of points surrounding the entire model.
+          // For example, if ASPECT's surface mesh is a 2D surface that is 3x3 (nx x ny) points,
+          // FastScape will be set as a 2D 5x5 point surface. On return to ASPECT, the outer ghost nodes
+          // will be ignored, and ASPECT will see only the inner 3x3 surface of FastScape.
+          if (use_ghost_nodes)
+            set_ghost_nodes(h.get(), vx.get(), vy.get(), vz.get());
+
+          // If specified, apply the orographic controls to the FastScape model.
+          if (use_orographic_controls)
+            apply_orographic_controls(h.get(), kd.get(), kf.get());
+
+          // Get current FastScape timestep.
+          int istep = 0;
+          fastscape_get_step_(&istep);
+
+          // Set velocity components.
+          if (use_velocities)
+            {
+              fastscape_set_u_(vz.get());
+              fastscape_set_v_(vx.get(), vy.get());
+            }
+
+          // Set h to new values, and erosional parameters if there have been changes.
+          fastscape_set_h_(h.get());
+          fastscape_set_erosional_parameters_(kf.get(), &kfsed, &m, &n, kd.get(), &kdsed, &g, &g, &p);
+
+          // Find  timestep size, run fastscape, and make visualizations.
+          execute_fastscape(h.get(), kd.get(), istep);
+
+          // Write a file to store h, b & step for restarting.
+          // TODO: It would be good to roll this into the general ASPECT checkpointing,
+          // and when we do this needs to be changed.
+          if (((this->get_parameters().checkpoint_time_secs == 0) &&
+               (this->get_parameters().checkpoint_steps > 0) &&
+               ((current_timestep + 1) % this->get_parameters().checkpoint_steps == 0)) ||
+              (this->get_time() + a_dt >= end_time && this->get_timestepping_manager().need_checkpoint_on_terminate()))
+            {
+              save_restart_files(h.get(), b.get(), sf.get(), istep);
+            }
+
+          // If we've reached the end time, destroy FastScape.
+          if (this->get_time() + this->get_timestep() > end_time)
```

Comment:
This will only work for the most common termination criteria (end time), but not for things like end step or steady state velocity. Can you move this into the destructor of this class? I am not sure using `this->pcout()` in the destructor is safe, but it is worth a try. Otherwise we could look into more complicated solutions, like asking the termination manager here, if the simulation will be executed after this time step. But this could have unintended consequences, as this operation changes the state of some termination criteria like the steady state computation. Let's try the destructor first. Do you know how to move this?

Reply:
I've moved it into MeshDeformationHandler<dim>::~MeshDeformationHandler () in interface.cc, is this correct? pcout did work here, but it probably isn't necessary so I've removed it.

Reply:
I meant to create a new destructor `Fastscape::~Fastscape`. That would be the best place to destroy any objects the Fastscape class has created.